### ISSUE TITLE
CNI getting multiple DEL calls for dead containers

### ISSUE THREAD

**Issue by caseydavenport** [kubernetes/kubernetes#44100]
**Kubernetes version** (use `kubectl version`):

```
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.0", GitCommit:"fff5156092b56e6bd60fff75aad4dc9de6b6ef37", GitTreeState:"clean", BuildDate:"2017-03-28T16:24:30Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

**What happened**:

I see multiple CNI DEL calls for stopped infra containers.

**What you expected to happen**:

Should only receive a single DEL call for the stopped infra container.

**How to reproduce it** (as minimally and precisely as possible):

Reboot a Kubernetes Node that is running at least one Pod.  When the node comes back up, CNI receives the following calls.

- DEL for old infra container.
- ADD for new infra container.
- DEL for old infra container.

**Anything else we need to know**:

This thread discussed how to prevent this issue when the GC behavior was added originally: https://github.com/kubernetes/kubernetes/pull/37036/files#r92748913

Seems like the fix in that PR isn't quite behaving as was expected.  I'm still trying to dig into what's causing the extra DEL call.

CC @dcbw

**Comment by yujuhong** [comment]
@caseydavenport kubelet has some extra safety mechanism that may cause extra network teardown calls. Did this cause any issue for you?

**Comment by caseydavenport** [comment]
@yujuhong yeah.  The Calico plugin stores some state keyed off the pod name/namespace, which means that when it gets the DEL for the old infra container _after_ the ADD, then it cleans up that state and that Pod stops working.

There is some code [here](https://github.com/dcbw/kubernetes/blob/20e1cdb97c7399225a5a1db623c77faade73527a/pkg/kubelet/dockertools/docker_manager.go#L1507) that tries to clean up any old network containers right before starting the new Pod, but it only tears down CNI and doesn't seem to delete any exited containers, which I think means the GC will call teardown again in the future.

**Comment by yujuhong** [comment]
> @yujuhong yeah. The Calico plugin stores some state keyed off the pod name/namespace, which means that when it gets the DEL for the old infra container after the ADD, then it cleans up that state and that Pod stops working.

The GC will call teardown before removing the infra container to handle cases where the network wasn't teardown properly. Shouldn't the teardown return ok in this case and the GC just move on...?

/cc @freehan 

> There is some code here that tries to clean up any old network containers right before starting the new Pod, but it only tears down CNI and doesn't seem to delete any exited containers, which I think means the GC will call teardown again in the future.

Just FYI, the code here is not used as the default in 1.6.

**Comment by caseydavenport** [comment]
> The GC will call teardown before removing the infra container to handle cases where the network wasn't teardown properly. Shouldn't the teardown return ok in this case and the GC just move on...?

The teardown does return OK.  I think the problem is that the teardown of the old infra container is occurring _after_ a new infra container has been networked for the Pod, which makes Calico think the Pod is being torn down when it's not.

> Just FYI, the code here is not used as the default in 1.6.

Ah, right.  Could you point me at the default code?

**Comment by yujuhong** [comment]
> The teardown does return OK. I think the problem is that the teardown of the old infra container is occurring after a new infra container has been networked for the Pod, which makes Calico think the Pod is being torn down when it's not.

Hmm...so the teardown doesn't use the infra container ID as a key?

We don't have any mechanism to distinguish whether the network has been torn down for a infra container. To err on the side of caution, we do perform extra teardown in many places. For example, the GC runs in a different thread and only look at the exited containers. If only the (name, namespace) tuple is used to identify a network, this can be an issue.

**Comment by caseydavenport** [comment]
> Hmm...so the teardown doesn't use the infra container ID as a key?

Nope.  This was originally done in Calico to work around the fact that older versions of Kubernetes would call "setup" multiple times for the same Pod but with different container IDs - it allowed the Calico plugin to track and cleanup resources for stale Pods without leaking.

> For example, the GC runs in a different thread and only look at the exited containers

Is there a reason that the infra containers aren't removed at the same time that the network is torn down?

The proposed fix in the original PR was to call CNI DEL on any exited infra containers for a Pod before calling ADD.  If in addition to that, the infra containers were removed before the ADD, then I think this issue would go away (since the GC would no longer need to remove them).  Does that make sense?

**Comment by yujuhong** [comment]
> Is there a reason that the exited infra containers aren't removed at the same time that the network is torn down?

In CRI (the new implementation), kubelet is only aware of the `PodSandbox` concept, which may be implemented as an infra container, a VM, or something else. If it was a VM, deleting the` PodSandbox` would end up deleting all the containers in it. kubelet can't cope with this just yet because 1) it uses the exited containers as checkpoints to know whether it should start a new one, and 2) in the current docker integration, the lifecycles of containers and its logs are tied together -- deleting the containers would mean we'd lose the logs forever. Both (1) and (2) can be resolved in the long run, but it takes time.

> Nope. This was originally done in Calico to work around the fact that older versions of Kubernetes would call "setup" multiple times for the same Pod but with different container IDs - it allowed the Calico plugin to track and cleanup resources for stale Pods without leaking.

I am not familiar with the implementation of the plugin and this setup issue you mentioned, but I think relying on just pod namespace and name is risky. If, for whatever reason, the pod was recreated (i.e., different UID but same name and namespace) and kubelet hasn't cleaned up the resource yet, you may encounter a more interesting race condition.

Is the setup issue still valid, and is it possible to change the plugin to use the container ID as a key?

**Comment by yujuhong** [comment]
@freehan I assigned this to you for now since you have more context in this area and your opinions are wanted :-)

**Comment by caseydavenport** [comment]
> If, for whatever reason, the pod was recreated (i.e., different UID but same name and namespace)

Yeah, this is exactly the scenario I'm hitting, and exactly the scenario that pushed us to originally pick name/namespace as the identifying information way back when.

> is it possible to change the plugin to use the container ID as a key?

It would be possible, though a bit unfortunate for compatibility / upgrade reasons.  If there's a way to make k8s handle this scenario that's not atrocious, that would be a much better solution for users, but understandable if that's not possible.

**Comment by freehan** [comment]
We only have 2 options here:
1. We guarantee that TearDownPod only happens successfully once before recreating the pod if needed.
2. Network plugins need to be tolerant to multiple TearDowns. 

There are multiple occasions where kubelet triggers StopPodSandbox(calls TearDownPod) for good reasons. They are in different go routines and happens asynchronously. 

Since StopPodSandbox is targeting pod sandbox instead of pod, I think it is reasonable for network plugins to distinguish resource by sandboxID instead of pod. Thoughts?

**Comment by caseydavenport** [comment]
The Calico plugin has been updated to handle this behavior as of [Calico CNI release v1.7.0](https://github.com/projectcalico/cni-plugin/releases/tag/v1.7.0).

> Since StopPodSandbox is targeting pod sandbox instead of pod, I think it is reasonable for network plugins to distinguish resource by podSandbox(use sandboxID) instead of pod. Thoughts?

I think this is reasonable enough for now, so I'll close this issue.

**Comment by timothysc** [comment]
@caseydavenport I believe this is the source of multiple race condition issues we have seen across multiple versions of calico.  

On some distro(s) e.g. fedora there is a kernel panic condition that I'm trying to record. 

On other environments it causes some race conditions on cluster bootstrap.  

I'm reopening this issue as a higher priority bug.  

/cc @sjenning @ncdc @derekwaynecarr

**Comment by caseydavenport** [comment]
@timothysc is there an issue somewhere for the specific failures and races you're seeing with Calico CNI v1.10.0?

Would be interested in seeing a log file to help diagnose.

**Comment by timothysc** [comment]
We believe https://github.com/projectcalico/calico/issues/1105 is symptomatic of this issue. 

We have also witnessed: https://paste.fedoraproject.org/paste/OXLjFPmS6L4u71WP4NS2ng which leads us to believe that a race condition still occurs on network namespace cleaning. 

An attempt to repro with level 5 logging and kernel dumping produces: https://drive.google.com/open?id=0ByIl30A0PqL5M1p4U3JIRzhWNUk

**Comment by sjenning** [comment]
@timothysc that kernel stack trace is fun.  Are you trying to mount an nfs share from a transient (non-host) network namespace?  Is this a containerized kubelet trying to do the mount?

It does seem like the the network namespace is being removed from underneath nfs even though there is reference counting on the net namespace struct.  Maybe we aren't taking the reference under lock or there is a double put() somewhere...

**Comment by timothysc** [comment]
> @timothysc that kernel stack trace is fun. Are you trying to mount an nfs share from a transient (non-host) network namespace? 

It's running the e2e's when it bombs, so I'm not entirely certain "what" it's doing.

> Is this a containerized kubelet trying to do the mount?

It's not containerized kubelet, it's kubeadm install of 1.7.5 kubelet on f26.  It's a single node cluster (master + worker)

**Comment by timothysc** [comment]
@sjenning here is the last flushed e2e.log for that data - https://drive.google.com/file/d/0ByIl30A0PqL5U2g4WW1iTjg5VE0/view?usp=sharing

**Comment by timothysc** [comment]
So there are two possible fixes here: 
1. De-dupe the flood of overlapping deletes that call through to CNI (Which is weird imo) 
2. Update calico-cni to handle this cleaner-er.  In the situation where we wedge a pod into a crashloop.

**Comment by dcbw** [comment]
@timothysc a while back I did https://github.com/kubernetes/kubernetes/pull/37036 which at least serialized the DEL calls so they don't overlap.  But this is how Kubelet currently works due to garbage collection, the sync loop, and clearing out containers.

We changed most CNI plugins to accept multiple DEL calls and be tolerant of missing namespaces on delete.

**Comment by dcbw** [comment]
@timothysc note that https://github.com/kubernetes/kubernetes/pull/37036 only serializes them in internal plugins.  If you use a remote runtime then you need to do that yourself, like I also did for CRIO in https://github.com/cri-o/ocicni/pull/5

**Comment by caseydavenport** [comment]
> We changed most CNI plugins to accept multiple DEL calls and be tolerant of missing namespaces on delete.

Yep, the Calico CNI plugin is also designed to be tolerant of a missing netns on delete. (see [this code](https://github.com/projectcalico/cni-plugin/blob/master/utils/utils.go#L49))

**Comment by dcbw** [comment]
@timothysc #37036 serializes calls on a per-pod basis, but it will (correctly IMHO) allow parallel operations for different pods.

**Comment by rantuttl** [comment]
I've seen this same issue, and put a theory in #51554. I somehow believe that most of the CNI failures that occur following the original successful delete are merely symptomatic of an original failure to delete the sandbox container.

**Comment by caseydavenport** [comment]
I've dug into this a bit (see [slack convo](https://kubernetes.slack.com/archives/C09QYUH5W/p1505436158000099))

The summary is that the Calico CNI plugin has code in place to handle the out-of-order ADD->DEL, and that code seems to be working in the logs above for most cases.

However, there is a race condition in which sometimes the Calico policy controller overwrites the container ID metadata (called "activeInstanceID") and thus the CNI plugin is unable to make that judgement call. This is outlined here: https://github.com/projectcalico/libcalico/issues/189

There's a fix for that in Calico `master` that will be in the next release.

You can try out the fix today by using the Calico master manifests, e.g. https://docs.projectcalico.org/master/getting-started/kubernetes/installation/hosted/kubeadm/1.6/calico.yaml

**Comment by rantuttl** [comment]
@caseydavenport @dcbw @timothysc -- I respect the possibility that you're seeing a slightly different problem, but in my case, and after reviewing my level 5 logs against the kubelet code, I am convinced that the problem lies with the kubelet not deleting the sandbox after asking the CNI to teardown the networking via kubelet's `TearDownPod` by way of `StopPodSandbox`.

In my logs, I can clearly see the CNI remove the sandbox networking, have IPAM release the IP, removing the entry from etcd, and return success back to the kubelet. It's at this point, I believe things go south. The kubelet then calls the docker-client to stop the containers in the pod by way of kubelet's `StopContainer` (also in `StopPodSandbox`). Docker does stop the sandbox container because I see `proc/<pid>/ns/net` go away from Linux OS, after which the pod's containers are "attempted" to be deleted. This is the important part, and explains why Exited sandbox containers are left hanging around.

kubelet's PLEG spots the stopped containers, and the kubelet `syncInterationLoop` picks this up and calls `cleanupContainersInPod`, which in-turn calls `getContainersToDeleteInPod` (by way of `deleteContainersInPod`). The kubelet code then checks `PodStatus` in `getContainersToDeleteInPod`. There are **two pointers** in `PodStatus`: one to `ContainerStatuses` and another to `SandboxStatuses`. In my logs, both of these had values within `PodStatus`. The output of this method are the containers to queue to a pod deleter worker. In the case of this code, only `ContainerStatuses` is checked, and therefore, the sandbox container is never queued for deletion.

Now, because the sandbox container is not deleted, I believe GC spots this and tries to delete this container. This results in continual calls CNI to teardown networking. However, it will always fail because the CNI already deleted the container, and the entry is not present in etcd. Furthermore, I would continually see the following nagging `nsenter` warning following this deletion attempt.

```
Aug 27 06:25:36 k8s-worker3 kubelet[22795]: W0827 06:25:36.703895   22795 docker_sandbox.go:263] NetworkPlugin cni failed on the status hook for pod "blue-im-alive-3284971222-1zjv6_blue-public": Unexpected command output nsenter: cannot open : No such file or directory
```

This is because subsequent calls to kubelet's `GetNetNS` always fail, and I continually see the above `nsenter` warning (the '--net=' is always an empty string passed to nsenter command because `proc/<pid>/ns/net` was previously deleted for this sandbox).

This is my view of my problem (and maybe this one too) based on my logs and review of the code. Still open to opinions on this theory.

**Comment by timothysc** [comment]
@rantuttl I believe there are a couple of separate issues here that are a result of non-deterministic behavior between kubelet <> host <> CNI plugins.   I think we've uncovered 1 in calico and @caseydavenport 's fix works for that issue.   However, I think the misbehavior of calico exposed 1 or two other bugs.  e.g. the kernel should never panic, and I think that may be linked to your issue.

**Comment by rantuttl** [comment]
@timothysc -- Yes, I suppose an issue in Calico may have been exposed. However, I did not see what @caseydavenport may be seeing, although I do see the continual calls to CNI, but never the kernel panic describe in Fedora (I use Ubuntu 16.04). I can surely provide my heavily annotated level 5 logs somewhere, which lead me to the conclusions I made above about my similar behaviors described in this issue.

btw, I am `rtuttle` on slack channel sig-networksig

**Comment by timothysc** [comment]
Please post a link to your data, it would be useful imo.

**Comment by rantuttl** [comment]
@timothysc -- As requested, my logs used to come to the conclusions noted above (sorry, Ubuntu pastebin)

http://paste.ubuntu.com/25540698/

**Comment by luxas** [comment]
@timothysc @dcbw @caseydavenport Any action items here for v1.9?

**Comment by timothysc** [comment]
@luxas we found/fixed a behavior issue in calico 2.6.  But this behavior that exists in the kubelet is still there and I'd argue it's pathological and should be addressed.

**Comment by rantuttl** [comment]
@luxas @timothysc I believe the Calico issue, and subsequent solution that was provided probably was a good solution for CNI being repeatedly invoked for DEL; however, and IMHO, it was only symptomatic of the root issue I described here previously on the kubelet. I believe it may be a very easy fix, although I have not coded it up to verify yet since I was looking for others' feedback on it.

**Comment by luxas** [comment]
@timothysc So you're saying the kubelet should NOT call DEL multiple times for dead containers?

**Comment by timchenxiaoyu** [comment]
firstly,why call CNI delete more than once, secondly ,why exit sanbox container no to be deleted?

**Comment by rantuttl** [comment]
Under the conditions I observed, no, kubelet should not be calling the CNI again. I actually observed that the sandbox container was removed from the network fabric. However, after returning back to kubelet, kubelet did NOT delete its view of the sandbox container (even though kubelet had already stopped the container via docker client, and the OS network namespace references had been removed from the underlying OS). Because it was still cached, kubelet just keeps calling the CNI over and over, never having success because it's already deleted via the CNI and docker client.

My earlier point (up a few comments) is that I can see there are two pointers in `PodStatus`: one to `ContainerStatuses` and another to `SandboxStatuses`. In my logs, both of these had values within PodStatus. The output of this method (`func getContainersToDeleteInPod`) are the containers to queue to a pod deleter worker. In the case of this code, only `ContainerStatuses` is checked; not `SandboxStatuses`.Therefore, the sandbox container is never queued for deletion.

**Comment by timchenxiaoyu** [comment]
if I upgrade calico cni to 1.11.1,the problem solve ,because calico cni return waring not error when sanbox deleting

**Comment by luxas** [comment]
@kubernetes/sig-node-bugs @kubernetes/sig-network-bugs any conclusion here?
This seems to be fixed on the Calico side and no other implementations seem to suffer badly by this suboptimal behavior, but is there actually someone that has cycles to fix this and would that in that case be a breaking change?
